sys/chunked_ringbuffer/chunked_ringbuffer.c

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (C) 2021 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ *
+ * @author  Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#include <string.h>
+#include "chunked_ringbuffer.h"
+#include "irq.h"
+
+static int _get_free_chunk(chunk_ringbuf_t *rb)
+{
+    int idx = rb->chunk_cur;
+    for (int i = 0; i < CHUNK_NUM_MAX; ++i) {
+        if (rb->chunk_start[idx] == NULL) {
+            return idx;
+        }
+
+        if (++idx == CHUNK_NUM_MAX) {
+            idx = 0;
+        }
+    }
+
+    return -1;
+}
+
+static int _get_complete_chunk(chunk_ringbuf_t *rb)
+{
+    int idx = rb->chunk_cur;
+    for (int i = 0; i < CHUNK_NUM_MAX; ++i) {
+        if (rb->chunk_start[idx] != NULL) {
```

Comment:
```suggestion
    int idx = atomic_load_u8(&rb->chunk_cur);
    for (int i = 0; i < CHUNK_NUM_MAX; ++i) {
        if (rb->chunk_start[idx] != NULL) {
```

Atomic for the reasons mentioned in the other comment (can't let the compiler reorder reads to earlier than the pointer access). I don't quite get when the loop would be used though, so I don't understand whether atomic access to the pointers would be necessary. (My gut feeling says not, but that also says the pointer can never be non-NULL).

Reply:
> I don't quite get when the loop would be used though

You are right, I think this could be written as

```suggestion
    if (rb->protect) {
        return rb->chunk_cur;
    }

    return -1;
```

Reply:
> a compiler (or CPU, but our CPUs don't do that) might reorder the actual buffer reads to happen before the index read that says whether data is here. 

I don't follow. The buffer pointer depends on the index - how would a compiler (or CPU) be allowed to re-order that?
`chunk_cur` is *only* set by `crb_consume_chunk()` so there isn't even any concurrency to worry about here.

Reply:
Chunk consumption *is* in a critical section (missed that) -- so right, that access is safe (and I should reread the full thing again).

However, *something* in a polling loop needs to be atomic (or in a critical section), for otherwise a `while !chunk_available() {}` loop has no point where the compiler needs to reload any memory, and the compiler is free to make a `if (!chunk_available()) { while (true) {} }` out of it (got chunk checking is pure). (It's undefined behavior even [in C++](https://stackoverflow.com/questions/41320725/is-infinite-loop-still-undefined-behavior-in-c-if-it-calls-shared-library)). (I think that touching the IRQ state qualifies, simply because it's assembly and thus out of the model).

Reply:
But isn't that only true if `chunk_available()` was `static`?
The compiler will consider a `bool chunk_available(void *ctx)` opaque, it can't optimize it away just because `ctx` didn't change between calls.

Reply:
I don't think that not being static helps against that, model-wise. At latest with `-flto`, the difference vanishes. AIU, static vs. non-static influences where things wind up on the object file level, but that's not part of the abstract machine.

Reply:
I suppose then we would have to do something like 

```patch
--- a/sys/chunked_ringbuffer/chunked_ringbuffer.c
+++ b/sys/chunked_ringbuffer/chunked_ringbuffer.
@@ -22,7 +23,8 @@ static int _get_free_chunk(chunk_ringbuf_t *rb)
 {
     int idx = rb->chunk_cur;
     for (int i = 0; i < CHUNK_NUM_MAX; ++i) {
-        if (rb->chunk_start[idx] == NULL) {
+        uint32_t _ptr = atomic_load_u32((void *)&rb->chunk_start[idx]);
+        if (_ptr == 0) {
             return idx;
         }
 
@@ -38,7 +40,8 @@ static int _get_complete_chunk(chunk_ringbuf_t *rb)
 {
     int idx = rb->chunk_cur;
     for (int i = 0; i < CHUNK_NUM_MAX; ++i) {
-        if (rb->chunk_start[idx] != NULL) {
+        uint32_t _ptr = atomic_load_u32((void *)&rb->chunk_start[idx]);
+        if (_ptr) {
             return idx;
         }

```

Reply:
Something like that. I'd hope that we manage to do all synchronization around the index (and not the pointers) to be portable more easily. (Otherwise we might need to extend the atomic_load family to atomic_load_ptr and concern ourselves with subtle differences between intptr_t and size_t).